File: src/main/java/by/sorface/sorface/gateway/config/handlers/StateRedirectUrlServerAuthenticationSuccessHandler.java
package by.sorface.sorface.gateway.config.handlers;

import org.springframework.security.core.Authentication;
import org.springframework.security.web.server.DefaultServerRedirectStrategy;
import org.springframework.security.web.server.WebFilterExchange;
import org.springframework.security.web.server.authentication.RedirectServerAuthenticationSuccessHandler;
import org.springframework.util.MultiValueMap;
import reactor.core.publisher.Mono;

import java.net.URI;
import java.util.Objects;

public class StateRedirectUrlServerAuthenticationSuccessHandler extends RedirectServerAuthenticationSuccessHandler {

    private static final String QUERY_STATE = "state";

    private static final String SEPARATOR_QUERY_PARAM_STATE = "~";

    @Override
    public Mono<Void> onAuthenticationSuccess(WebFilterExchange webFilterExchange, Authentication authentication) {
        return Mono.defer(() -> {
            final MultiValueMap<String, String> queryParams = webFilterExchange.getExchange().getRequest().getQueryParams();

            boolean hasState = queryParams.containsKey(QUERY_STATE);

            if (hasState) {
                final var state = queryParams.getFirst(QUERY_STATE);

                if (Objects.isNull(state)) {
                    return super.onAuthenticationSuccess(webFilterExchange, authentication);
                }

                final var split = state.split(SEPARATOR_QUERY_PARAM_STATE);

                if (split.length <= 1) {
                    return super.onAuthenticationSuccess(webFilterExchange, authentication);
                }

                final String redirectUrl = split[1];

                return new DefaultServerRedirectStrategy().sendRedirect(webFilterExchange.getExchange(), URI.create(redirectUrl));
            }

            return super.onAuthenticationSuccess(webFilterExchange, authentication);
        });
    }
}

File: src/main/kotlin/by/sorface/gateway/config/handlers/StateRedirectUrlServerAuthenticationSuccessHandler.kt
package by.sorface.gateway.config.handlers

import by.sorface.gateway.config.constants.WebSessionAttributes
import org.springframework.security.core.Authentication
import org.springframework.security.web.server.DefaultServerRedirectStrategy
import org.springframework.security.web.server.WebFilterExchange
import org.springframework.security.web.server.authentication.RedirectServerAuthenticationSuccessHandler
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebSession
import reactor.core.publisher.Mono
import java.net.URI
import java.net.URISyntaxException

/**
 * Обработчик успешного входа в систему
 */
class StateRedirectUrlServerAuthenticationSuccessHandler : RedirectServerAuthenticationSuccessHandler() {

    companion object {
        private val DEFAULT_SERVER_REDIRECT_STRATEGY = DefaultServerRedirectStrategy()
    }

    init {
        DEFAULT_SERVER_REDIRECT_STRATEGY.setContextRelative(false)
    }

    /**
     * Обработчик успешной авторизации
     */
    override fun onAuthenticationSuccess(webFilterExchange: WebFilterExchange, authentication: Authentication): Mono<Void> {
        return getRedirectUrl(webFilterExchange.exchange)
            .switchIfEmpty(Mono.defer { super.onAuthenticationSuccess(webFilterExchange, authentication).then(Mono.empty()) })
            .flatMap { redirectUrl: URI? -> DEFAULT_SERVER_REDIRECT_STRATEGY.sendRedirect(webFilterExchange.exchange, redirectUrl) }
    }

    /**
     * Получение из сессии URL для перенаправления
     *
     * @param exchange исходный запрос
     */
    private fun getRedirectUrl(exchange: ServerWebExchange): Mono<URI> {
        return exchange.session
            .filter { webSession: WebSession -> webSession.attributes.containsKey(WebSessionAttributes.ORIGINAL_REQUEST_ATTRIBUTE) }
            .flatMap { webSession: WebSession ->
                val redirectUrl = webSession.attributes[WebSessionAttributes.ORIGINAL_REQUEST_ATTRIBUTE] as String? ?: return@flatMap Mono.empty<URI>()

                webSession.attributes.remove(WebSessionAttributes.ORIGINAL_REQUEST_ATTRIBUTE)

                try {
                    return@flatMap Mono.just<URI>(URI(redirectUrl))
                } catch (e: URISyntaxException) {
                    return@flatMap Mono.error<URI>(e)
                }
            }
            .onErrorResume { Mono.empty() }
    }

}
